File: token-validation-core/src/main/java/no/nav/security/token/support/core/context/TokenValidationContext.java
package no.nav.security.token.support.core.context;

import no.nav.security.token.support.core.jwt.JwtToken;
import no.nav.security.token.support.core.jwt.JwtTokenClaims;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class TokenValidationContext {

    private final Map<String, JwtToken> issuerShortNameValidatedTokenMap;

    public TokenValidationContext(Map<String, JwtToken> issuerShortNameValidatedTokenMap) {
        this.issuerShortNameValidatedTokenMap = issuerShortNameValidatedTokenMap;
    }

    public Optional<JwtToken> getJwtTokenAsOptional(String issuerName) {
        return jwtToken(issuerName);
    }

    public Optional<JwtToken> getFirstValidToken() {
        return issuerShortNameValidatedTokenMap.values().stream().findFirst();
    }

    public
    JwtToken getJwtToken(String issuerName) {
        return jwtToken(issuerName).orElse(null);
    }

    @Override
    public String toString() {
        return "TokenValidationContext{" +
            "issuers=" + issuerShortNameValidatedTokenMap.keySet() +
            '}';
    }

    public
    JwtTokenClaims getClaims(String issuerName) {
        return jwtToken(issuerName)
            .map(JwtToken::getJwtTokenClaims)
            .orElse(null);
    }

    public Optional<JwtTokenClaims> getAnyValidClaims() {
        return issuerShortNameValidatedTokenMap.values().stream()
            .map(JwtToken::getJwtTokenClaims)
            .findFirst();
    }

    public boolean hasValidToken() {
        return !issuerShortNameValidatedTokenMap.isEmpty();
    }

    public boolean hasTokenFor(String issuerName) {
        return jwtToken(issuerName).isPresent();
    }

    public List<String> getIssuers() {
        return new ArrayList<>(issuerShortNameValidatedTokenMap.keySet());
    }

    private Optional<JwtToken> jwtToken(String issuerName) {
        return issuerShortNameValidatedTokenMap.containsKey(issuerName) ?
            Optional.of(issuerShortNameValidatedTokenMap.get(issuerName))
            : Optional.empty();
    }
}

File: token-validation-core/src/main/kotlin/no/nav/security/token/support/core/context/TokenValidationContext.kt
package no.nav.security.token.support.core.context

import java.util.Optional
import no.nav.security.token.support.core.jwt.JwtToken

class TokenValidationContext(private val validatedTokens : Map<String, JwtToken>) {

    fun getJwtTokenAsOptional(issuerName : String) = jwtToken(issuerName)?.let { Optional.of(it) } ?: Optional.empty()

    val firstValidToken get() = validatedTokens.values.firstOrNull()
    fun getJwtToken(issuerName : String) = jwtToken(issuerName)

    fun getClaims(issuerName : String) = jwtToken(issuerName)?.jwtTokenClaims ?: throw IllegalArgumentException("No token found for issuer $issuerName")

    val anyValidClaims get() =
        Optional.ofNullable(validatedTokens.values
            .map(JwtToken::jwtTokenClaims)
            .firstOrNull())

    fun hasValidToken()  = validatedTokens.isNotEmpty()

    fun hasTokenFor(issuerName : String) = getJwtToken(issuerName) != null
    val issuers get() = validatedTokens.keys.toList()

    private fun jwtToken(issuerName: String) = validatedTokens[issuerName]

    override fun toString() = "TokenValidationContext{issuers=${validatedTokens.keys}}"

}